webTables

IlyaKirdun opened the pull request:
Добавлено: types.ts , registrationForm.page.ts , webTables.page.ts , webTables.spec.ts(13 cases) .
Обновлено: functions.ts .

9 for 'Web Tables' page:
1. Проверка функционала на создание пользователя.
2. Проверка функционала поиска в таблице.
3. Проверка функционала сортировки в таблице.
4. Проверка функционала редактирования пользователя.
5. Проверка функционала удаления пользователя в таблице.
6. Проверка отображения индикации при поиске несуществующего пользователя.
7. Проверка функционала пагинации в таблице.
8. Проверка функционала перехода страницы по номеру в таблице.
9. Проверка функционала выбора количества строк в таблице.

6 for 'Registration Form' in modal window:
1. Проверка функционала открытия и закрытия.
2. Проверка функционала на заполнение полей ввода:
    First Name, Last Name, Email, Age, Salary, Department.
3. Проверка функционала на некорректный ввод.
4. Проверка функционала на смену индикации некорректного ввода.
5. Проверка функционала на некорректный ввод в модальном окне.
6. Проверка функционала на смену индикации некорректного ввода в модальном окне.

IlyaKirdun committed:
Удалил неактуальные методы и локаторы в webTables.page.ts,
сделал метод fillPageNumberInput приватным,
импортировал метод assertByState и подставил взамен удалённых,
исправлены опечатки и неточности в webTables.spec.ts,
Изменил логику в методе checkAmountRows

IlyaKirdun committed:
Изменил названия методов\переменных,
переработал метод navigateToMainPage,
интегрировал кастомный тип в navigateToCard(),
Добавил массив departmentList,
интегрировал кастомный тип в метод getUserData(),
улучшена логика в методе fillPageNumberInput(),
переработал\добавил методы webTables.page.ts,
переработал тест 'CASE_11: Проверка функционала пагинации в таблице.',

IlyaKirdun committed:
Переработал webTables.spec.ts ,

Рефактор navigationBar.ts :
Добавлен объект navigationBarStructure,
переработал метод clickElementInNavigationBar(),

Рефактор registrationModalWindow.ts:
Добавлен объект colorMap,
переписал локаторы,
удалил неактуальные методы,
переработал\добавил методы,
интегрировал кастомные типы в методы,

IlyaKirdun committed:
Рефактор functions.ts:
переработал\добавил методы,
исправление опечаток.

Добавил types.ts,
добавил кастомные типы в types.ts,

Рефактор webTables.page.ts:
Изменение назваий в методах\переменных,
переработал метод deleteAddedUsers(),

Прописали baseUrl в playwright.config.ts,
добавил исключение в .gitignore.

AliaksandrPuchok commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "demoqa",
-  "version": "1.0.1",
+  "version": "1.1.2",
```
поправим после сквоша

AliaksandrPuchok commented on page/main.page.ts:
```diff
@@ -7,11 +8,11 @@ export default class MainPage {
 
-  async gotoMainPage(): Promise<void> {
+  async navigateToMainPage(): Promise<void> {
     await this.page.goto('https://demoqa.com', { waitUntil: 'domcontentloaded' })
```
это baseUrl надо настроить в конфиге

AliaksandrPuchok commented on page/webTables.page.ts:
```diff
@@ -49,37 +60,38 @@ export default class WebTablesPage {
-    expect(expectUserData).not.toBe(currentUserData)
+  public async fillPageNumberInput(pageNumber: number): Promise<void> {
+    await this.fill(this.searchInput, `${pageNumber}`)
+    await this.page.keyboard.press('Enter')
```
разве searchInput и pageNumberInput одинаковые?

AliaksandrPuchok replied:
не увидел ответа

AliaksandrPuchok commented on page/webTables.page.ts:
```diff
@@ -49,37 +60,38 @@ export default class WebTablesPage {
-  async fillSearchInput(userData: string): Promise<void> {
-    await this.searchInput.fill(userData)
+  private async fill(nameInput: Locator, fillData: string): Promise<void> {
+    await nameInput.fill(fillData)
```
мб вместо nameInput стоит назвать inputLocator?

rybignat replied:
Тоже зарезолвил неисправленное

AliaksandrPuchok commented on page/webTables.page.ts:
```diff
@@ -93,55 +105,64 @@ export default class WebTablesPage {
-  async isTableEmpty(): Promise<void> {
+  async verifyIndicatorTableEmpty(): Promise<void> {
     await expect(this.noRowsInput).toBeVisible()
   }
```
Мб лучше checkIndicatorTableEmptyState и добавить state toBeVisible() & toBeHidden() с базовым значением toBeHidden

и проверять можно будет в начале тестов

rybignat replied:
А почему резолв,если метод не переделан?

AliaksandrPuchok commented on page/webTables.page.ts:
```diff
@@ -93,55 +105,64 @@ export default class WebTablesPage {
-  async checkAmountRows(row: number): Promise<void> {
-    const getCurrentAmountRows = await this.page.$$(`//div[@class="rt-tbody"]//div[@class="rt-tr-group"]`)
-    expect(getCurrentAmountRows).toHaveLength(row)
+  async deleteAddedUsers(userCount: number, usersEmail: string[]): Promise<void> {
+    for (let index: number = 0; index < userCount; index++) {
+      await this.clickDeleteUserButton(usersEmail[index])
+    }
   }
```
почему не упростить до 
```
async deleteAddedUsers(usersEmail: string[]): Promise<void> {
  for (const email of usersEmail) {
    await this.clickDeleteUserButton(email)
  }
}
```

AliaksandrPuchok commented on utils/components/navigationBar.ts:
```diff
@@ -1,15 +1,32 @@
+    let parentElementName: string | undefined = await getParentNameByChildName(this.navigationBarStructure, childElementName)
+
+  if(typeof parentElementName !== 'string' ){
+    process.exit(1)
```
Лучше выбрасывать ошибку чтобы ты понимал причину падения
```
  throw new Error(`Не найден родительский элемент для "${childElementName}"`)
```

AliaksandrPuchok commented on utils/components/registrationModalWindow.ts:
```diff
@@ -1,65 +1,58 @@
-  async isModalWindowToBeVisible(): Promise<void> {
-    await expect(this.modalWindow).toBeVisible()
-  }
+  async checkVisibilityRegistrationForm(state: 'toBeVisible'|'toBeHidden'): Promise<void> {
```
checkRegistrationFormVisibility

AliaksandrPuchok commented on utils/components/registrationModalWindow.ts:
```diff
@@ -1,65 +1,58 @@
-  async fillInputInModalWindow(inputName: string, testUserData: string): Promise<void> {
-    await this.page.locator(`//div[@class= "modal-body"]
-    //input[@placeholder= "${inputName}"]`).fill(`${testUserData}`)
+  async fillInputInModalWindow(inputName: SortingCell, testUserData: string): Promise<void> {
```
fillModalWindowInput

AliaksandrPuchok commented on utils/components/registrationModalWindow.ts:
```diff
@@ -1,65 +1,58 @@
   }
 
-  async isInputDataCorrectInModalWindow(inputName: string, testUserData: string): Promise<void> {
+  async verifyInputDataCorrectInModalWindow(inputName: SortingCell, testUserData: string): Promise<void> {
```
verifyEnteredDataInModal

AliaksandrPuchok commented on utils/components/registrationModalWindow.ts:
```diff
@@ -1,65 +1,58 @@
-    let stateInput: Locator = this.page.locator(`//div[@class= "modal-body"]//input[@placeholder= "${inputName}"]`)
-    const backgroundColor = await stateInput.evaluate(el => getComputedStyle(el).borderColor)
-    expect(stateValid[state]).toBe(backgroundColor)
+  async verifyInputColorInModalWindow(inputName: SortingCell, state: 'valid' | 'invalid'): Promise<void> {
```
verifyModalWindowInputColor

AliaksandrPuchok commented on utils/functions.ts:
```diff
@@ -1,18 +1,45 @@
+  assertion.toBe(expected);
+}
+
+export async function getParentNameByChildName<T> (obj: {[key: string]: T[]}, value: T): Promise<string | undefined> {
```
не понимаю зачем функция асинхронная

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -137,15 +139,40 @@ test.describe('Проверка функциональности на стран
-      await new Promise(resolve => setTimeout(resolve, 1000))
-      await registrationModalWindow.validationInputColorInModalWindow('First Name', 'invalidColor')
+    await test.step('Проверяем индикацию поля "Email" при некорректном вводе.', async () => {
+      await new Promise(resolve => setTimeout(resolve, 500))
```
мне кажется что задержки лучше вынести в методы перед которыми они нужны, сократит код, упростит поддержку

AliaksandrPuchok replied:
не вижу выноса в метод

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -370,214 +397,264 @@ test.describe('Проверка функциональности на стран
+
+    await test.step(`Post-condition`, async () => {
+      await test.step('Удаляем пользователя', async () => {
+        await webTablesPage.clickDeleteUserButton(testDataForCreateUser['userEmail'])
```
необходимо уточнить что clickDeleteUserButtonByUserEmail

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -370,214 +397,264 @@ test.describe('Проверка функциональности на стран
 
       await test.step('Заготавливаем пользователей.', async () => {
-        await registrationModalWindow.usersGeneration()
+        usersEmail = await webTablesPage.addUsersGenerator(10)
```
не usersEmail, а userEmails

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -370,214 +397,264 @@ test.describe('Проверка функциональности на стран
-    await test.step('Сохраняем номер текущей страницы.', async () => {
-      pageBeforePagination = await webTablesPage.getPageNumber()
+    await test.step('Сохраняем номер текущей страницы и пользователя из первой строки.', async () => {
+      pageBeforePagination = await webTablesPage.pageNumberInput.inputValue()
```
может initialPage, а не pageBeforePagination

AliaksandrPuchok reviewed (changes requested):
нужны правки по оставленным комментам

rybignat commented on page/webTables.page.ts:
```diff
@@ -93,55 +105,64 @@ export default class WebTablesPage {
-    let currentPage: string | null = await this.pageNumberInput.getAttribute('Value')
-    expect(beforePage).toBe(currentPage)
+  async checkAmountRows(row: SelectingAmountRows): Promise<void> {
+    const getCurrentAmountRows = this.page.locator(`//div[@class="rt-tbody"]//div[@class="rt-tr-group"]`)
```
почему у тебя константа проименована по типу метода? Больше подходит rowsAmount или rowsQuantity

rybignat commented on page/webTables.page.ts:
```diff
@@ -93,55 +105,64 @@ export default class WebTablesPage {
-  async fillPageNumbers(pageNumber: string): Promise<void> {
-    await this.pageNumberInput.fill(pageNumber)
-    await this.clickOnTable.click()
+  async addUsersGenerator(userCount: number): Promise<string[]> {
```
Тут уже определись либо addUsers или usersGenerator

rybignat commented on page/webTables.page.ts:
```diff
@@ -93,55 +105,64 @@ export default class WebTablesPage {
+  }
 
-    return pageNumber
+  async selectRowsOnPage(manyRows: number): Promise<void> {
```
скорее rowsPerPage и не manyRows а просто rows

rybignat reviewed (changes requested):
Чутка поправить и в принципе пойдет. Работай над неймингом.
Вот неплохой гайд, https://ru.hexlet.io/blog/posts/rukovodstvo-po-imenovaniyu-klassov-funktsiy-i-peremennyh  но можно посмотреть и на других ресурсах.

AliaksandrPuchok commented on page/main.page.ts:
```diff
@@ -7,11 +8,11 @@ export default class MainPage {
 
-  async gotoCard(keyWordFromLocator: string): Promise<void> {
-    await this.page.locator(`//h5[text()="${keyWordFromLocator}"]`).click()
+  async navigateToCard(elementName: ElementsOnMainPage): Promise<void> {
```
это ведь клик, а не переход

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -370,214 +397,268 @@ test.describe('Проверка функциональности на стран
+    await test.step('Pre-condition.', async () => {
       await test.step('Выбрать отображение в "5 rows"', async () => {
-        await webTablesPage.selectRowsOnPage('5')
+        await webTablesPage.selectRowsOnPage(5)
```
почему данные захардкожены а не вынесены в начало теста?

AliaksandrPuchok commented on tests/webTables.spec.ts:
```diff
@@ -370,214 +397,268 @@ test.describe('Проверка функциональности на стран
-      await test.step('Выбрать отображение в "5 rows"', async () => {
-        await webTablesPage.selectRowsOnPage('5')
+      await test.step(`Выбрать отображение в ${initialAmountRows}.`, async () => {
+        await webTablesPage.selectRowsOnPage(5)
```
в тест степе обращаешься к переменной а почему в аргументы не передашь?

AliaksandrPuchok reviewed (changes requested):
будь внимательнее к тому что написано в комментариях

IlyaKirdun committed:
Изменение названий в методах\переменных,
исправление опечаток,
переработал тест 'CASE_13: Проверка функционала выбора количества строк в таблице.' в webTables.spec.ts
интегрировал кастомный тип

IlyaKirdun committed:
Изменение названия метода clickEditUserButton(),
удален неактуальный локатор,
исправление опечатки,
добавление переменных в тест.

IlyaKirdun commented:
Вроде все косяки закрыл и 2 раза проверил все ли изменения внес в коммит)

rybignat commented on page/webTables.page.ts:
```diff
@@ -49,37 +58,38 @@ export default class WebTablesPage {
 
-  async fillSearchInput(userData: string): Promise<void> {
-    await this.searchInput.fill(userData)
+  private async fill(nameLocator: Locator, fillData: string): Promise<void> {
```
nameLocator - локатор имени
locatorName - имя локатора

IlyaKirdun replied:
ууу сука, а как же забить хуй. ведь ключевое слово лучше писать в конце?

rybignat replied:
У тебя ключевое слово это name

rybignat commented on page/webTables.page.ts:
```diff
@@ -93,55 +103,64 @@ export default class WebTablesPage {
-    await this.page.locator(`//div[@class="rt-tr-group"][1]//div[@class="rt-td"][7]//span[@title= "Edit"]`).click()
+  async clickEditUserButtonByUserEmail(userEmail: string): Promise<void> {
+    await this.page.locator(`//div[@class="rt-tr-group"]//div[text()="${userEmail}"]/parent::div//span[@title="Edit"]`).click()
   }
 
-  async deleteUser(): Promise<void> {
-    await this.page.locator(`//div[@class="rt-tr-group"][1]//div[@class="rt-td"][7]//span[@title= "Delete"]`).click()
+  async clickDeleteUserButtonByUserEmail(userEmail: string): Promise<void> {
+    await this.page.locator(`//div[@class="rt-tr-group"]//div[text()="${userEmail}"]/parent::div//span[@title="Delete"]`).click()
   }
```
У тебя в локаторе отличие в одно слово. 
Можно объединить метод под названием clickUsersActionButtonByUserEmail

IlyaKirdun replied:
Ты прав

rybignat commented on page/webTables.page.ts:
```diff
@@ -93,55 +103,64 @@ export default class WebTablesPage {
-  async isPageMatch(beforePage: string): Promise<void> {
-    let currentPage: string | null = await this.pageNumberInput.getAttribute('Value')
-    expect(beforePage).toBe(currentPage)
+  async checkAmountRows(row: SelectingAmountRows): Promise<void> {
```
checkRowsAmount

IlyaKirdun replied:
Та же шляпа что и с локатором, жду ответ)

rybignat commented on tests/textBox.spec.ts:
```diff
@@ -63,7 +63,7 @@ test.describe('Check the functioning of the "Text Box" section', () => {
 
     await test.step(`Match the "Permanent Address" output field and ${testData.permanentAddress}`, async () => {
-      await textBoxPage.isDataMatch(`Permanent Address :${testData.permanentAddress}`, textBoxPage.permanentAddressOutput)
+      await textBoxPage.isDataMatch(`Permananet Address :${testData.permanentAddress}`, textBoxPage.permanentAddressOutput)
```
опечатка

IlyaKirdun replied:
нет, беда демоки. Именно такая хуйня приходит с сайта

rybignat commented on tests/webTables.spec.ts:
```diff
@@ -370,218 +384,278 @@ test.describe('Проверка функциональности на стран
+      await assertByState(actualUser, initialUserEmail, 'notMatch')
+    })
+
+    await test.step('Post-condition', async () => {
+      await test.step('Удаляем созданных пользователей', async () => {
+        await webTablesPage.deleteAddedUsers(userEmails)
+      })
```
Добавить постконды с удалением созданных пользователей в остальные кейсы

IlyaKirdun replied:
Вообще в каждый или ты про 1 тест кейс где я не удаляю одного пользователя?

rybignat replied:
В каждый где ты создаешь пользователя

rybignat reviewed (changes requested):
внимательно просмотреть 8 кейс и отредактировать

IlyaKirdun committed:
переработал тест 'CASE_8: Проверка функционала редактирования пользователя'
